File: python/Q91_decode_ways.py
#
# @lc app=leetcode id=91 lang=python3
#
# [91] Decode Ways
#

# @lc code=start
class Solution:
    def numDecodings(self, s: str) -> int:
        if len(s) == 1:
            return 0 if s == "0" else 1

        dp: list[int] = [0] * len(s)
        dp[0] = 0 if s[0] == "0" else 1
        if s[1] != "0":
            dp[1] += dp[0]
        if 10 <= int(s[0:2]) <= 26:
            dp[1] += 1

        for i in range(2, len(s)):
            char: str = s[i]
            previous_one: bool = s[i - 1] == "1"
            previous_two: bool = s[i - 1] == "2"

            match char:
                case "0":
                    if not (previous_one or previous_two):
                        return 0

                    dp[i] = dp[i - 2]

                case "1" | "2" | "3" | "4" | "5" | "6":
                    if previous_one or previous_two:
                        dp[i] = dp[i - 1] + dp[i - 2]
                    else:
                        dp[i] = dp[i - 1]

                case "7" | "8" | "9":
                    if previous_one:
                        dp[i] = dp[i - 1] + dp[i - 2]
                    else:
                        dp[i] = dp[i - 1]

        return dp[-1]


# @lc code=end

File: rust/src/leetcode/Q91_decode_ways.rs
/*
 * @lc app=leetcode id=91 lang=rust
 *
 * [91] Decode Ways
 */

// @lc code=start
impl Solution {
    pub fn num_decodings(s: String) -> i32 {
        let chars: Vec<char> = s.chars().collect();
        let n: usize = chars.len();

        if n == 1 {
            return if chars[0] == '0' { 0 } else { 1 };
        }

        let mut dp: Vec<i32> = vec![0; n];
        dp[0] = if chars[0] == '0' { 0 } else { 1 };

        if chars[1] != '0' {
            dp[1] += dp[0];
        }

        let first_two: i32 = format!("{}{}", chars[0], chars[1]).parse::<i32>().unwrap();
        if (10..=26).contains(&first_two) {
            dp[1] += 1;
        }

        for i in 2..n {
            let char = chars[i];
            let previous_one: bool = chars[i - 1] == '1';
            let previous_two: bool = chars[i - 1] == '2';

            match char {
                '0' => {
                    if !(previous_one || previous_two) {
                        return 0;
                    }
                    dp[i] = dp[i - 2];
                }
                '1' | '2' | '3' | '4' | '5' | '6' => {
                    if previous_one || previous_two {
                        dp[i] = dp[i - 1] + dp[i - 2];
                    } else {
                        dp[i] = dp[i - 1];
                    }
                }
                '7' | '8' | '9' => {
                    if previous_one {
                        dp[i] = dp[i - 1] + dp[i - 2];
                    } else {
                        dp[i] = dp[i - 1];
                    }
                }
                _ => return 0,
            }
        }

        dp[n - 1]
    }
}
// @lc code=end
